aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+    n_chunks = math.ceil(len(objs) / chunk_size)
+    total_size_mb = sys.getsizeof(objs) / (1024 * 1024)
+    if total_size_mb > 10:
+        # bigger than 10MB
+        if (total_size_mb / n_chunks) > 6:
+            chunk_size = math.floor(total_size_mb / 6)
+            logger.info(
+                'The size of the objects to insert into BQ is very large, '
+                f'adjusting the chunk size to {chunk_size}'
+            )
+
+    if n_chunks > 1:
+        logger.info(f'Will insert {len(objs)} rows in {n_chunks} chunks')
+
+    inserts = 0
+
+    for chunk_idx, chunked_objs in enumerate(chunk(objs, chunk_size)):
+
+        _query = f"""
+            SELECT id FROM `{table}`
+            WHERE id IN UNNEST(@ids);
+        """
+
+        ids = set(o['id'] for o in chunked_objs)
+        if len(ids) != len(chunked_objs):
+            counter = defaultdict(int)
+            for o in chunked_objs:
+                counter[o['id']] += 1
+            duplicates = [f'{k} ({v})' for k, v in counter.items() if v > 1]
+            raise ValueError(
+                'There are multiple rows with the same id: ' + ', '.join(duplicates)
+            )
+
+        job_config = bq.QueryJobConfig(
+            query_parameters=[
+                bq.ArrayQueryParameter('ids', 'STRING', list(ids)),
+            ]
+        )
+
+        result = get_bigquery_client().query(_query, job_config=job_config).result()
+        existing_ids = set(result.to_dataframe()['id'])
+
+        # Filter out any rows that are already in the table
+        filtered_obj = [o for o in chunked_objs if o['id'] not in existing_ids]
+
+        nrows = len(filtered_obj)
+
+        if nrows == 0:
+            logger.info(f'Not inserting any rows (0/{len(chunked_objs)})')
+            continue
+
+        if dry_run:
+            logger.info(f'DRY_RUN: Inserting {nrows}/{len(chunked_objs)} rows')
+            inserts += nrows
+            continue
+
+        # Count number of rows adding
+        logger.info(
+            f'Inserting {nrows}/{len(chunked_objs)} rows '
+            f'({chunk_idx+1}/{n_chunks} chunk)'
+        )
+
+        # Insert the new rows
+        job_config = bq.LoadJobConfig()
+        job_config.source_format = bq.SourceFormat.NEWLINE_DELIMITED_JSON
+        job_config.schema = get_formatted_bq_schema()
+
+        j = '\n'.join(json.dumps(o) for o in filtered_obj)
+
+        resp = get_bigquery_client().load_table_from_file(
+            StringIO(j), table, job_config=job_config
+        )
+        try:
+            result = resp.result()
+            logger.info(
+                f'Inserted {result.output_rows}/{nrows} rows '
+                f'({chunk_idx+1}/{n_chunks} chunk)'
+            )
+        except ClientError as e:
+            logger.error(resp.errors)
+            raise e
+
+        inserts += nrows
+
+    return inserts
+
+
+def insert_dataframe_rows_in_table(table: str, df: pd.DataFrame):
+    """Insert new rows from dataframe into a table"""
```

Comment:
Which table? What schema is expected? It's pretty unclear for a caller when to use this without reading the code.

Reply:
Refactored to make it a bit clearer. General data import function to import any df rows into the gcp table. The provided schema, and df schema must match the table.